### DISCUSSION TITLE
Why is `createJSObjectReference` not usable for functions?

### DISCUSSION THREAD

**Discussion by trungnt2910** [dotnet/aspnetcore#55259] [Q&A] [answered]
`createJSObjectReference`, the function that powers `IJSRuntime` and `IJSObjectReference`, has this line:

https://github.com/dotnet/aspnetcore/blob/7033ec7f402fa4b7df2bf72a5e3dff1df6831f57/src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts#L148

This means things like this will fail:

```cs
        await using var func = await JSRuntime.InvokeAsync<IJSObjectReference>("window.alert");
        await func.InvokeVoidAsync("call", "Hello World!");
```

Why is that check there in the first place? Without such restrictions, issues like https://github.com/dotnet/aspnetcore/issues/31151 are basically solved.

**Comment by oroztocil** [comment, marked as answer]
This has been implemented in https://github.com/dotnet/aspnetcore/pull/61453 which relaxes the condition to:

```csharp
if (jsObject && (typeof jsObject === "object" || jsObject instanceof Function)) { ... }
```

**Reply by trungnt2910** [reply]
Interesting to see this actually getting read by the team. Thanks!

Now we can access most JavaScript runtime functionality from `JSRuntime` without requiring any actual helper JS code:
https://trungnt2910.com/blazor-goodbye-js#attempt---reflectconstruct

```cshtml
@inject IJSRuntime JSRuntime

<button @onclick="Button_Click">Click me</button>

@code {
    private async Task Button_Click()
    {
        await using var ctor = await JSRuntime.InvokeAsync<IJSObjectReference>("Set.bind");
        await using var set = await JSRuntime.InvokeAsync<IJSObjectReference>("Reflect.construct",
            new object?[] { ctor, new string[] { "lxmonika" } });
        await JSRuntime.InvokeVoidAsync("console.log", set);
    }

}
```

**Reply by trungnt2910** [reply]
Actually just saw the new `InvokeNewAsync` API, which makes my trick obsolete 😅
